orttraining/orttraining/python/training/ortmodule/_io.py

```diff
@@ -305,12 +334,17 @@ def _add_dynamic_shape(name, input):
         return dynamic_axes
 
     def _add_input(name, input, onnx_graph, onnx_graph_input_names):
-        if input is not None and (onnx_graph is None or name in onnx_graph_input_names):
-            if input.requires_grad:
-                input_names_require_grad.append(name)
+        if input is not None:
+            if (onnx_graph is None or name in onnx_graph_input_names):
+                if isinstance(input, torch.Tensor):
+                    if input.requires_grad:
+                        input_names_require_grad.append(name)
+                    dynamic_axes.update(_add_dynamic_shape(name, input))
+                    input_shape.append(list(input.size()))
+
+            # InputInfo should contain all the names irrespective of whether they are
```

Comment:
Double check this part. I remember using the fact that input_names only contains *used* inputs on my PRs regarding to kwargs (the last 2)

Review those PRs and make sure this change do not break anything that I might have failed in writing UTs for

Reply:
This was actually a subtle bug we introduced earlier. Here is the illustration of the bug:

- Imagine we have a model with a boolean input:
```py
class PrimitiveTypesInputNet(torch.nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(PrimitiveTypesInputNet, self).__init__()

        self.fc1 = torch.nn.Linear(input_size, hidden_size)
        self.relu = torch.nn.ReLU()
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, input1, bool_argument, int_argument, float_argument):
        input1 = input1 + int_argument + float_argument
        if bool_argument:
            out = self.fc1(input1)
            out = self.relu(out)
            out = self.fc2(out)
        else:
            out = self.fc1(input1)
            out = self.fc2(out)
            out = self.relu(out)
        return out
```
- On export, the bool input will be dropped in the onnx model.
- We would also have the ```bool_argument``` not a part of the ```input_info.names```. So, ```input_info.names``` would be ```[input1, int_argument, float_argument]```
- The inputs defined in the onnx model also would be ```[input1, int_argument, float_argument]```
- When the user provides inputs like ```(torch.tensor([1, 2, 3], False, 123, 1.234)``` to the forward method and since we only iterate over the onnx model inputs, we would see that the the the ```int_argument``` corresponds to the 2nd position in the input list, we would pick up the value of False, for the ```int_argument``` which is incorrect.

Basically, the ```input_info``` must contain all the non ```None``` inputs to the forward method, regardless of whether the input belongs to the onnx graph.